---
title: "Viral Genomes Analysis"
author: "Francesco Lescai and Thomas Bleazard"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: readable
    highlight: tango
    toc: true
    toc_float: true
    css: $baseDir/assets/nibsc_report.css
editor_options:
  chunk_output_type: console
params:
  vcf: NULL
  callers: NULL
  samples: NULL
  genome: NULL
  genemodel: NULL
  baseDir: NULL
  bamSamples: NULL
  bamFiles: NULL
---


# Introduction

The analysis has been generating with the following steps:

TO BE COMPLETED

# Results

The following results are reported by sample, and have been generated with the procedure described above.

## Read Based Variant Analysis

```{r parseInput, include=FALSE}
library(tidyverse)
vcfFiles <- strsplit(params$vcf, ",")
samples <- strsplit(params$samples, ",")
callers <- strsplit(params$callers, ",")

vcfInfo <- data.frame(
  sample = samples,
  caller = callers,
  vcf = vcfFiles,
  stringsAsFactors = FALSE
)
names(vcfInfo) <- c("sample", "caller", "vcf")

basedir <- params$basedir

bamSamples <- strsplit(params$bamSamples, ",")
bamFiles <- strsplit(params$bamFiles, ",")

bamInfo <- data.frame(
  sample = bamSamples,
  bam = bamFiles,
  stringsAsFactors = FALSE
)
names(bamInfo) <- c("sample", "bam")

sampleData <- vcfInfo %>%
  left_join(bamInfo, by = "sample")
```


```{r run-summary-md, include=FALSE}
library(Gviz)
library(VariantAnnotation)
library(GenomicFeatures)
library(rtracklayer)
library(Biostrings)
library(tidyverse)
library(knitr)
out = NULL
for (index in 1:dim(sampleData)[[1]]) {
  genome <- params$genome
  variants <- file
  model <- params$genemodel
  sample <- sampleData[index,]$sample
  caller <- sampleData[index,]$caller
  vcffile <- sampleData[index,]$vcf
  bamfile <- sampleData[index,]$bam
  env = new.env()
  out = c(out, knit_child(paste0(baseDir, "/docs/loop_sample_variants.Rmd"), envir=env))
}

```

`r paste(out, collapse = '\n')`
